File: implementations/data structures/linked lists/circular/circular_s_linked_list.cpp
//
//  main.cpp
//  circular_d_linked_list
//
//  Created by Lareen Melo on 1/27/20.
//  Copyright © 2020 Lareen Melo. All rights reserved.
//

#include <iostream>
#include <assert.h>

template <class T>
struct Node {
    T value;
    Node *next;
};

template <class T>
class CircularLinkedList {
    Node<T> *head;
    Node<T> *tail;
    
public:
    CircularLinkedList() {
        head = nullptr;
        tail = nullptr;
    }
    
    void append(T value) {
        Node<T> *new_node = new Node<T>;
        
        new_node->value = value;
        new_node->next = head;
        
        if (head == nullptr) {
            head = new_node;
            tail = new_node;
            new_node->next = head;
            
        } else {
            Node<T> *current = tail;
            
            current->next = new_node;
            tail = new_node;
            
        }
    }
    
    void remove() {
        assert(head != nullptr);
        
        if (head == tail) {
            head = nullptr;
            tail = nullptr;
            
        } else {
            Node<T> *previous = nullptr;
            Node<T> *current = head;

            
            while (current->next != head) {
                previous = current;
                current = current->next;
            }
            current->next = nullptr;
            previous->next = head;
            
            tail = previous;
            
            delete current;
            current = nullptr;
            
            
        }
    }

    
    // prints list references, last element's reference is first
    void print() {
        Node<T> *current = head;
        
        if (current == nullptr) {
            std::cout << "emtpy list" << std::endl;
        }

        do {
            if (current->next == head) {
                std::cout << current->value << std::endl;
            } else {
                std::cout << current->value << "->";
            }
            current = current->next;

        } while (current != head);
    }
};

int main() {
    CircularLinkedList<int> circular_list;
    circular_list.append(0);
    circular_list.append(1);
    circular_list.append(2);
    circular_list.append(3);
    circular_list.append(4);
    circular_list.append(5);

    circular_list.remove();
    circular_list.remove();
    circular_list.remove();
    circular_list.remove();
    circular_list.remove();
    circular_list.remove();

    circular_list.print();
    return 0;
}

File: implementations/data structures/linked lists/circular/circular_s_linked_list.swift
//
//  main.swift
//  circular_s_linked_list
//
//  Created by Lareen Melo on 1/27/20.
//  Copyright © 2020 Lareen Melo. All rights reserved.
//

import Foundation

class Node<T: Equatable>: Equatable {
    public static func == (lhs: Node<T>, rhs: Node<T>) -> Bool {
        return lhs.value == rhs.value && lhs.next === rhs.next
    }
    
    var value: T
    var next: Node<T>? = nil
    
    init(with value: T) {
        self.value = value
    }
}


class LinkedList<T: Equatable> {
    private var head: Node<T>?
    private var tail: Node<T>?
    
    func append(_ value: T) {
        let newValue = Node(with: value)
        let currentNode = tail
        
        if head == nil {
            head = newValue
            tail = newValue
            
            newValue.next = head

        } else {
            currentNode?.next = newValue
            newValue.next = head
            tail = newValue

        }
        
    }
    
    func remove() {
        assert(head != nil)
        
        if head == tail {
            head = nil
            tail = nil
        } else {
            var predecessorNode = head
            
            while predecessorNode?.next != tail {
                predecessorNode = predecessorNode?.next
            }
            
            var currentNode = predecessorNode?.next
            predecessorNode?.next = currentNode?.next
            currentNode?.next = nil
            tail = predecessorNode
            currentNode = nil
            
        }
    }
    
    func prettyPrint() {
        var currentNode = head
        var list = ""
        
        repeat {
            if currentNode?.next == head {
                list += "\(currentNode!.value)->\(currentNode!.next!.value)"
                
            } else {
                list += "\(currentNode!.value)->"
            }
            
            currentNode = currentNode?.next
            
        } while currentNode != head
        
        print(list)
    }
}


var list = LinkedList<Int>()

list.append(1)
list.append(2)
list.append(3)
list.append(4)
list.append(6)

list.remove()
list.append(6)

list.prettyPrint()
